### ISSUE TITLE
BUG: Cannot install SciPy using pip on Raspberry Pi

### ISSUE THREAD

**Issue by rekkitcwts** [scipy/scipy#16784]
### Describe your issue.

Tried to install scipy through pip3 install -3 requirements.txt with the following:
`librosa==0.6.3
numpy
pandas
soundfile==0.9.0
wave
scikit-learn==0.24.2
tqdm==4.28.1
matplotlib==2.2.3
pyaudio==0.2.11`
but it gave an error. Already tried using sudo apt-get install python3-scipy, but still the same issue happens.

### Reproducing Code Example

```python
N/A - all I did is run sudo LLVM_CONFIG=`which llvm-config-11` pip3 install -r requirements.txt
```

### Error message

```shell
Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [55 lines of output]
      The Meson build system
      Version: 0.62.2
      Source dir: /tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821
      Build dir: /tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821/.mesonpy-b9th5ncm/build
      Build type: native build
      Project name: SciPy
      Project version: 1.9.0
      C compiler for the host machine: cc (gcc 8.3.0 "cc (Raspbian 8.3.0-6+rpi1) 8.3.0")
      C linker for the host machine: cc ld.bfd 2.31.1
      C++ compiler for the host machine: c++ (gcc 8.3.0 "c++ (Raspbian 8.3.0-6+rpi1) 8.3.0")
      C++ linker for the host machine: c++ ld.bfd 2.31.1
      Host machine cpu family: arm
      Host machine cpu: armv7l
      Compiler for C supports arguments -Wno-unused-but-set-variable: YES
      Library m found: YES
      Fortran compiler for the host machine: gfortran (gcc 8.3.0 "GNU Fortran (Raspbian 8.3.0-6+rpi1) 8.3.0")
      Fortran linker for the host machine: gfortran ld.bfd 2.31.1
      Program cython found: YES (/tmp/pip-build-env-wco6m20p/overlay/bin/cython)
      Program pythran found: YES (/tmp/pip-build-env-wco6m20p/overlay/bin/pythran)
      Program cp found: YES (/usr/bin/cp)
      Program python3 found: YES (/usr/local/bin/python3.9)
      Found pkg-config: /usr/bin/pkg-config (0.29)
      
      ../../scipy/meson.build:40:0: ERROR: Command "/usr/local/bin/python3.9 -c import os; os.chdir(".."); import numpy; print(numpy.get_include())" failed with status 1.
      
      A full log can be found at /tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821/.mesonpy-b9th5ncm/build/meson-logs/meson-log.txt
      + meson setup --native-file=/tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821/.mesonpy-native-file.ini -Ddebug=false -Doptimization=2 --prefix=/usr/local /tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821 /tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821/.mesonpy-b9th5ncm/build
      Traceback (most recent call last):
        File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
          main()
        File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 130, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-wco6m20p/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 909, in get_requires_for_build_wheel
          with _project(config_settings) as project:
        File "/usr/local/lib/python3.9/contextlib.py", line 119, in __enter__
          return next(self.gen)
        File "/tmp/pip-build-env-wco6m20p/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 888, in _project
          with Project.with_temp_working_dir(
        File "/usr/local/lib/python3.9/contextlib.py", line 119, in __enter__
          return next(self.gen)
        File "/tmp/pip-build-env-wco6m20p/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 547, in with_temp_working_dir
          yield cls(source_dir, tmpdir, build_dir)
        File "/tmp/pip-build-env-wco6m20p/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 463, in __init__
          self._configure(reconfigure=bool(build_dir) and not native_file_mismatch)
        File "/tmp/pip-build-env-wco6m20p/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 494, in _configure
          self._meson(
        File "/tmp/pip-build-env-wco6m20p/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 477, in _meson
          return self._proc('meson', *args)
        File "/tmp/pip-build-env-wco6m20p/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 472, in _proc
          subprocess.check_call(list(args))
        File "/usr/local/lib/python3.9/subprocess.py", line 373, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['meson', 'setup', '--native-file=/tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821/.mesonpy-native-file.ini', '-Ddebug=false', '-Doptimization=2', '--prefix=/usr/local', '/tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821', '/tmp/pip-install-93y57wnw/scipy_34185744f91f4e619985e97d31e27821/.mesonpy-b9th5ncm/build']' returned non-zero exit status 1.
      [end of output]
```

### SciPy/NumPy/Python version information

Python 3.9 - NumPy 1.22.4

**Comment by tupui** [comment]
Hi @rekkitcwts, thank you for reporting. If I am reading this correctly it's because it needs to build a wheel and to do so, NumPy needs to be present before as it's a build dependency. So try installing NumPy before doing that.

**Comment by rgommers** [comment]
This is weird. `numpy` should be installed when `pip` is invoked, because it's a build-time dependency in `pyproject.toml`.

@rekkitcwts would you mind re-running that with `-vv` added to the `pip install` command and then providing the complete build log (maybe put it in a [gist](https://gist.github.com/ if it's large)?

**Comment by dancaron** [comment]
@rekkitcwts Try the following:

Tested on: OSX 11.1. Python 3.9.5. M1 native. Scipy 1.9.0.

### Install openblas
```brew install openblas```

### Install additional dependencies
```
pip install cython pybind11 pythran
pip install numpy
```

### Get paths that are required to be set in environment
```brew info openblas```

```
openblas: stable 0.3.20 (bottled), HEAD [keg-only]
...
For pkg-config to find openblas you may need to set:
  export PKG_CONFIG_PATH="/opt/homebrew/opt/openblas/lib/pkgconfig"
```

### Execute the above export statement in your terminal
But get your own path from `brew info openblas`

```
pip install scipy
```

**Comment by tupui** [comment]
@dancaron Please downgrade openblas to 0.3.18 as the last 2 versions have known issues on M1 (fixes in the pipeline for conda-forge).

**Comment by tylerjereddy** [comment]
Just a thought, but looking at that `meson` log, this is a Raspberry Pi, right? The compiler toolchain seems to suggest that, and it would be consistent with a bunch of source build failures because compatible binaries may not be available on default PyPI.

If my guess is right, you might try something like https://www.raspberrypi.com/news/piwheels/ perhaps? To check maybe do `uname -m` and let us know what it says? I think Pi boards use something other than the ARMv8/aarch64 that we support/provide binaries for? @rgommers or @h-vetinari may know better than I.

**Comment by rekkitcwts** [comment]
> Just a thought, but looking at that `meson` log, this is a Raspberry Pi, right? The compiler toolchain seems to suggest that, and it would be consistent with a bunch of source build failures because compatible binaries may not be available on default PyPI.
> 
> If my guess is right, you might try something like https://www.raspberrypi.com/news/piwheels/ perhaps? To check maybe do `uname -m` and let us know what it says? I think Pi boards use something other than the ARMv8/aarch64 that we support/provide binaries for? @rgommers or @h-vetinari may know better than I.

Yep -- it's a raspberry pi 3b+ with Buster in it -- and it fixed itself when I downgraded the Python from 3.9 to 3.7 due to some "glibc" errors.

**Comment by yumemio** [comment]
This is still happening on `scipy >= 1.9.0` and a Raspberry Pi 4B, even using wheels from www.piwheels.org.

A workaround is downgrading to Scipy 1.8.1, with `pip install scipy==1.8.1`.

Could you reopen this issue, or should I open another?

**Comment by rgommers** [comment]
This seems to not have been resolved conclusively for Raspberry Pi, so let's reopen this one. @yumemio please add the full build log for what you are seeing, and if it's due to `openblas` missing, how you installed that yourself.

**Comment by tupui** [comment]
I could try to debug if needed. I have a pi4.

**Comment by yumemio** [comment]
@rgommers @tupui Thanks for your response. Here is the minimal sequence of commands that produces the error on a Raspberry Pi 4 (32-bit Raspbian).

```console
# apt-get update
# apt-get install --yes \
    python3-pip \
    gfortran \
    libopenblas-base \
    libopenblas-dev
$ python3 -m pip install scipy
```

Here is the log from `python3 -m pip install scipy`. It automatically tries installing Scipy 1.9.3, 1.9.2, ..., 1.9.0 in turn, which all fail with an error. The installation succeeds only after falling back to `scipy-1.8.1`.

```
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting scipy
  Downloading scipy-1.9.3.tar.gz (42.1 MB)
     |????????????????????????????????| 42.1 MB 46 kB/s
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... error
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 /tmp/tmpzbi4vcl9_in_process.py prepare_metadata_for_build_wheel /tmp/tmpeewvhj2v
         cwd: /tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f
    Complete output (71 lines):
    The Meson build system
    Version: 0.64.1
    Source dir: /tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f
    Build dir: /tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f/.mesonpy-u88p7q3f/build
    Build type: native build
    Project name: SciPy
    Project version: 1.9.3
    C compiler for the host machine: cc (gcc 10.2.1 "cc (Raspbian 10.2.1-6+rpi1) 10.2.1 20210110")
    C linker for the host machine: cc ld.bfd 2.35.2
    C++ compiler for the host machine: c++ (gcc 10.2.1 "c++ (Raspbian 10.2.1-6+rpi1) 10.2.1 20210110")
    C++ linker for the host machine: c++ ld.bfd 2.35.2
    Host machine cpu family: arm
    Host machine cpu: armv7l
    Compiler for C supports arguments -Wno-unused-but-set-variable: YES
    Compiler for C supports arguments -Wno-unused-but-set-variable: YES (cached)
    Compiler for C supports arguments -Wno-unused-function: YES
    Compiler for C supports arguments -Wno-conversion: YES
    Compiler for C supports arguments -Wno-misleading-indentation: YES
    Compiler for C supports arguments -Wno-incompatible-pointer-types: YES
    Library m found: YES
    Fortran compiler for the host machine: gfortran (gcc 10.2.1 "GNU Fortran (Raspbian 10.2.1-6+rpi1) 10.2.1 20210110")
    Fortran linker for the host machine: gfortran ld.bfd 2.35.2
    Compiler for Fortran supports arguments -Wno-conversion: YES
    Program cython found: YES (/tmp/pip-build-env-qgj4s8jd/overlay/bin/cython)
    Program pythran found: YES (/tmp/pip-build-env-qgj4s8jd/overlay/bin/pythran)
    Program cp found: YES (/usr/bin/cp)
    Program python found: YES (/usr/bin/python3)
    Found pkg-config: /usr/bin/pkg-config (0.29.2)
    Run-time dependency threads found: YES

    ../../scipy/meson.build:40:0: ERROR: Command "/usr/bin/python3 -c import os; os.chdir(".."); import numpy; print(numpy.get_include())" failed with status 1.

    A full log can be found at /tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f/.mesonpy-u88p7q3f/build/meson-logs/meson-log.txt
    + meson setup --prefix=/usr /tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f /tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f/.mesonpy-u88p7q3f/build --native-file=/tmp/pip-install-3okqoo15/scipy_a5af$
2fae16c4250875f20d9db43706f/.mesonpy-native-file.ini -Ddebug=false -Doptimization=2
    Traceback (most recent call last):
      File "/tmp/tmpzbi4vcl9_in_process.py", line 126, in prepare_metadata_for_build_wheel
        hook = backend.prepare_metadata_for_build_wheel
    AttributeError: module 'mesonpy' has no attribute 'prepare_metadata_for_build_wheel'

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/tmp/tmpzbi4vcl9_in_process.py", line 280, in <module>
        main()
      File "/tmp/tmpzbi4vcl9_in_process.py", line 263, in main
        json_out['return_val'] = hook(**hook_input['kwargs'])
      File "/tmp/tmpzbi4vcl9_in_process.py", line 130, in prepare_metadata_for_build_wheel
        return _get_wheel_metadata_from_wheel(backend, metadata_directory,
      File "/tmp/tmpzbi4vcl9_in_process.py", line 159, in _get_wheel_metadata_from_wheel
        whl_basename = backend.build_wheel(metadata_directory, config_settings)
      File "/tmp/pip-build-env-qgj4s8jd/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 1060, in build_wheel
        with _project(config_settings) as project:
      File "/usr/lib/python3.9/contextlib.py", line 117, in __enter__
        return next(self.gen)
      File "/tmp/pip-build-env-qgj4s8jd/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 975, in _project
        with Project.with_temp_working_dir(
      File "/usr/lib/python3.9/contextlib.py", line 117, in __enter__
        return next(self.gen)
      File "/tmp/pip-build-env-qgj4s8jd/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 750, in with_temp_working_dir
        yield cls(source_dir, tmpdir, build_dir, meson_args)
      File "/tmp/pip-build-env-qgj4s8jd/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 632, in __init__
        self._configure(reconfigure=bool(build_dir) and not native_file_mismatch)
      File "/tmp/pip-build-env-qgj4s8jd/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 680, in _configure
        self._meson('setup', *setup_args)
      File "/tmp/pip-build-env-qgj4s8jd/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 657, in _meson
        return self._proc('meson', *args)
      File "/tmp/pip-build-env-qgj4s8jd/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 652, in _proc
        subprocess.check_call(list(args), env=self._env)
      File "/usr/lib/python3.9/subprocess.py", line 373, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['meson', 'setup', '--prefix=/usr', '/tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f', '/tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f/.mesonpy-u88p7q3f/build
', '--native-file=/tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f/.mesonpy-native-file.ini', '-Ddebug=false', '-Doptimization=2']' returned non-zero exit status 1.
    ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/0a/2e/44795c6398e24e45fa0bb61c3e98de1cfea567b1b51efd3751e2f7ff9720/scipy-1.9.3.tar.gz#sha256=fbc5c05c85c1a02be77b1ff591087c83bc44579c6d2bd9fb798bb64ea5e1a027 (from https://pypi.
org/simple/scipy/) (requires-python:>=3.8). Command errored out with exit status 1: /usr/bin/python3 /tmp/tmpzbi4vcl9_in_process.py prepare_metadata_for_build_wheel /tmp/tmpeewvhj2v Check the logs for full command output.
  Downloading scipy-1.9.2.tar.gz (42.1 MB)
     |????????????????????????????????| 42.1 MB 12 kB/s
...

(truncated; same error messages for scipy-1.9.{0..2})

...
WARNING: Discarding https://files.pythonhosted.org/packages/a8/e3/4ec401f609d34162b7023a09165da491630879e4cfa2336667fe2102cd06/scipy-1.9.0.tar.gz#sha256=c0dfd7d2429452e7e94904c6a3af63cbaa3cf51b348bd9d35b42db7e9ad42791 (from https://pypi.
org/simple/scipy/) (requires-python:>=3.8,<3.12). Command errored out with exit status 1: /usr/bin/python3 /tmp/tmpwg0zswlm_in_process.py get_requires_for_build_wheel /tmp/tmpeq5a_4j3 Check the logs for full command output.
  Downloading https://www.piwheels.org/simple/scipy/SciPy-1.8.1-cp39-cp39-linux_armv7l.whl (72.2 MB)
     |????????????????????????????????| 72.2 MB 3.1 kB/s
Requirement already satisfied: numpy<1.25.0,>=1.17.3 in /usr/lib/python3/dist-packages (from scipy) (1.19.5)
Installing collected packages: scipy
Successfully installed scipy-1.8.1
```

A more verbose log with `python3 -m pip uninstall scipy && python3 -m pip install -vv scipy` is available [here](https://gist.github.com/yumemio/f944d91876ce09271871018773894b9b).

**Comment by rgommers** [comment]
Thanks @rekkitcwts. This is the relevant part:
```
    ../../scipy/meson.build:40:0: ERROR: Command "/usr/bin/python3 -c import os; os.chdir(".."); import numpy; print(numpy.get_include())" failed with status 1.

    A full log can be found at /tmp/pip-install-3okqoo15/scipy_a5afe2fae16c4250875f20d9db43706f/.mesonpy-u88p7q3f/build/meson-logs/meson-log.txt
```
It's not clear why that fails, given that `numpy` must be installed already in the build env. The answer to that will be in `meson-log.txt`. `pip` likely has cleaned that up already, which is annoying. If so, using `--no-clean` may help.

> Here is the log from `python3 -m pip install scipy`. It automatically tries installing Scipy 1.9.3, 1.9.2, ..., 1.9.0 in turn, which all fail with an error.

That rebuilding behavior has been fixed in a more recent version of `pip` than `20.3.4` (not sure exactly which version). It should now just stop after encountering a build error.

**Comment by yumemio** [comment]
@rgommers, thanks again for your comment. Unfortunately, `meson-log.txt` is not found after `pip install`ing SciPy, even with the `--no-clean` option. My wild guess is that [Meson ate the directory](https://github.com/mesonbuild/meson/issues/8707), but I don't know if that is the case.

```console
$ python3 -m pip uninstall scipy && python3 -m pip install --no-clean scipy==1.9.3

(truncated)

    ../../scipy/meson.build:40:0: ERROR: Command "/usr/bin/python3 -c import os; os.chdir(".."); import numpy; print(numpy.get_include())" failed with status 1.

    A full log can be found at /tmp/pip-install-t5dmudk7/scipy_b81a8f2778594ae4974200269d0ea2f2/.mesonpy-nggmk0ia/build/meson-logs/meson-log.txt
    + meson setup --prefix=/usr /tmp/pip-install-t5dmudk7/scipy_b81a8f2778594ae4974200269d0ea2f2 /tmp/pip-install-t5dmudk7/scipy_b81a8f2778594ae4974200269d0ea2f2/.meson
py-nggmk0ia/build --native-file=/tmp/pip-install-t5dmudk7/scipy_b81a8f2778594ae4974200269d0ea2f2/.mesonpy-native-file.ini -Ddebug=false -Doptimization=2

(truncated)

$ ls /tmp/pip-install-t5dmudk7/scipy_b81a8f2778594ae4974200269d0ea2f2/.mesonpy-nggmk0ia/build/meson-logs/meson-log.txt
ls: cannot access '/tmp/pip-install-t5dmudk7/scipy_b81a8f2778594ae4974200269d0ea2f2/.mesonpy-nggmk0ia/build/meson-logs/meson-log.txt': No such file or directory

$ ls -Fa /tmp/pip-install-t5dmudk7/scipy_b81a8f2778594ae4974200269d0ea2f2
./            CONTRIBUTING.rst      do.py                 LICENSE.txt               mypy.ini               pytest.ini    setup.py*
../           dev.py                HACKING.rst.txt       meson.build               mypy_requirements.txt  README.rst    site.cfg.example
benchmarks/   doc/                  INSTALL.rst.txt       meson_options.txt         PKG-INFO               runtests.py*  tools/
CITATION.bib  doc_requirements.txt  LICENSES_bundled.txt  .mesonpy-native-file.ini  pyproject.toml         scipy/        tox.ini

$ cd /tmp/pip-install-t5dmudk7/scipy_b81a8f2778594ae4974200269d0ea2f2/
$ find . -regextype posix-egrep -iname '*log*' '!' -iregex '.*\.(c|h|cpp|hpp|ipp)$'
./doc/source/tutorial/examples/morphology_binary_dilation_erosion.py
./doc/source/tutorial/stats/continuous_loggamma.rst
./doc/source/tutorial/stats/continuous_halflogistic.rst
./doc/source/tutorial/stats/continuous_genlogistic.rst
./doc/source/tutorial/stats/continuous_genhalflogistic.rst
./doc/source/tutorial/stats/continuous_logistic.rst
./doc/source/tutorial/stats/discrete_logser.rst
./doc/source/tutorial/stats/continuous_powerlognorm.rst
./doc/source/tutorial/stats/continuous_loguniform.rst
./doc/source/tutorial/stats/continuous_loglaplace.rst
./doc/source/tutorial/stats/continuous_lognorm.rst
./doc/source/_static/scipy-mathjax/extensions/HelpDialog.js
./doc/source/_static/logo.svg
./tools/write_release_and_log.py
./tools/validate_runtests_log.py
./scipy/ndimage/tests/test_morphology.py
./scipy/ndimage/morphology.py
./scipy/ndimage/_morphology.py
./scipy/special/cdflib/rlog1.f
./scipy/special/cdflib/rlog.f
./scipy/special/_loggamma.pxd
./scipy/special/tests/data/boost/log1p_expm1_data_ipp
./scipy/special/tests/data/boost/log1p_expm1_data_ipp/log1p_expm1_data.txt
./scipy/special/tests/data/local/log1p_expm1_complex.txt
./scipy/special/tests/test_logit.py
./scipy/special/tests/test_log_softmax.py
./scipy/special/tests/test_loggamma.py
./scipy/special/tests/test_logsumexp.py
./scipy/special/amos/zlog.f
./scipy/special/_xlogy.pxd
./scipy/special/_logsumexp.py
./scipy/special/_precompute/loggamma.py
./scipy/_lib/boost/boost/preprocessor/logical
./scipy/_lib/boost/boost/log
./scipy/_lib/boost/boost/logic
./scipy/_lib/unuran/UNURAN_ChangeLog
./scipy/fft/_fftlog_multimethods.py
./scipy/fft/tests/test_fftlog.py
./scipy/fft/_fftlog.py
./scipy/sparse/linalg/_propack/PROPACK/Changelog
./scipy/io/matlab/tests/data/logical_sparse.mat
./scipy/io/matlab/tests/test_pathological.py
./benchmarks/benchmarks/linalg_logm.py
```

**Comment by rgommers** [comment]
Thanks @yumemio, that detailed output helped clear something up. Pip's `--no-clean` leaves behind the build environment, but that doesn't help us here. The SciPy build itself, done by `meson-python`, uses a `tmpdir` as well and it gets  cleaned up. What we need instead of `--no-clean` is to explicitly specify the build directory to use. This works for me:
```
FC=crashme pip install . --config-settings=builddir=build
```
That leaves behind a local `build/` dir with the needed `meson-log.txt` in it. Could you try that?

**Comment by yumemio** [comment]
Thanks @rgommers, and sorry for replying late. The Meson log revealed the actual issue: `libcblas.so.3` was missing.

```console
$ pip install --config-settings=builddir=/home/me/scipy/build scipy==1.9.3
...
$ cat scipy/build/meson-logs/meson-log.txt
...
(truncated)
...
Run-time dependency threads found: YES
Running command: /usr/bin/python3 -c import os; os.chdir(".."); import numpy; print(numpy.get_include())
--- stdout ---

--- stderr ---
Traceback (most recent call last):
  File "/tmp/pip-build-env-y7tf5n1e/overlay/lib/python3.9/site-packages/numpy/core/__init__.py", line 22, in <module>
    from . import multiarray
  File "/tmp/pip-build-env-y7tf5n1e/overlay/lib/python3.9/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/tmp/pip-build-env-y7tf5n1e/overlay/lib/python3.9/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ImportError: libcblas.so.3: cannot open shared object file: No such file or directory

...
(truncated)
...
```

I have uploaded the complete log without truncation [here](https://gist.github.com/yumemio/f442c7b572a278c837150540e26b7ec5) for reference.

After running `apt install libatlas-base-dev` (as described in [this StackOverflow answer](https://stackoverflow.com/a/56972113/13301046)), `scipy==1.9.3` was installed successfully.

Therefore the complete installation process for RPi4 would be:

```console
# apt install gfortran cmake libopenblas-base libopenblas-dev libatlas-base-dev
$ pip install scipy==1.9.3
```

**Comment by rgommers** [comment]
Ah great, that explains it. Thanks for following up @yumemio. So I think we're good here, all that's needed is to improve the stdout output on build failures, which is https://github.com/mesonbuild/meson-python/issues/246. So I'll close this again.

**Comment by sparky4** [comment]
i got this particular issue with Debian 12 python 3.11

**Comment by mdufrane** [comment]
@sparky4 I've got the exact same error, do you know how to fix it ? Thanks

**Comment by spearman** [comment]
> i got this particular issue with Debian 12 python 3.11
> 
> ```
>       [788/1572] Compiling C object scipy/stats/_unuran/unuran_wrapper.cpython-311-x86_64-linux-gnu.so.p/.._..__lib_unuran_unuran_src_distr_cvec.c.o
>       [789/1572] Compiling C object scipy/stats/_unuran/unuran_wrapper.cpython-311-x86_64-linux-gnu.so.p/.._..__lib_unuran_unuran_src_distr_cxtrans.c.o
>       [790/1572] Generating 'scipy/stats/_qmc_cy.cpython-311-x86_64-linux-gnu.so.p/_qmc_cy.cpp'.
>       [791/1572] Generating scipy/stats/_hypotests_pythran with a custom command
>       FAILED: scipy/stats/_hypotests_pythran.cpp
>       /tmp/pip-build-env-9_w94buz/overlay/bin/pythran -E ../../scipy/stats/_hypotests_pythran.py -o scipy/stats/_hypotests_pythran.cpp
>       Traceback (most recent call last):
>         File "/tmp/pip-build-env-9_w94buz/overlay/bin/pythran", line 8, in <module>
>           sys.exit(run())
>                    ^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/run.py", line 180, in run
>           pythran.compile_pythranfile(args.input_file,
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/toolchain.py", line 477, in compile_pythranfile
>           output_file = compile_pythrancode(module_name, fd.read(),
>                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/toolchain.py", line 395, in compile_pythrancode
>           module, error_checker = generate_cxx(module_name, pythrancode, specs, opts,
>                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/toolchain.py", line 138, in generate_cxx
>           pm, ir, docstrings = front_middle_end(module_name, code, optimizations,
>                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/toolchain.py", line 105, in front_middle_end
>           refine(pm, ir, optimizations, report_times)
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/middlend.py", line 63, in refine
>           apply_optimisation |= pm.apply(optimization, node, run_times)[0]
>                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/passmanager.py", line 243, in apply
>           ret=a.apply(node)
>               ^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/passmanager.py", line 201, in apply
>           new_node = self.run(node)
>                      ^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/passmanager.py", line 189, in run
>           n = super(Transformation, self).run(node)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/passmanager.py", line 97, in run
>           return self.visit(node)
>                  ^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/pythran/optimizations/pattern_transform.py", line 327, in visit
>           matcher = pattern()
>                     ^^^^^^^^^
>       TypeError: 'module' object is not callable
>       [792/1572] Compiling C object scipy/stats/_unuran/unuran_wrapper.cpython-311-x86_64-linux-gnu.so.p/.._..__lib_unuran_unuran_src_distr_distr.c.o
>       [793/1572] Compiling C object scipy/stats/_unuran/unuran_wrapper.cpython-311-x86_64-linux-gnu.so.p/.._..__lib_unuran_unuran_src_distr_discr.c.o
>       [794/1572] Compiling C object scipy/stats/_levy_stable/levyst.cpython-311-x86_64-linux-gnu.so.p/meson-generated_levyst.c.o
>       [795/1572] Compiling C++ object scipy/stats/_biasedurn.cpython-311-x86_64-linux-gnu.so.p/meson-generated__biasedurn.cpp.o
>       [796/1572] Generating 'scipy/stats/_unuran/unuran_wrapper.cpython-311-x86_64-linux-gnu.so.p/unuran_wrapper.c'.
>       [797/1572] Compiling C object scipy/stats/_sobol.cpython-311-x86_64-linux-gnu.so.p/meson-generated__sobol.c.o
>       [798/1572] Compiling C++ object scipy/stats/_qmc_cy.cpython-311-x86_64-linux-gnu.so.p/meson-generated__qmc_cy.cpp.o
>       [799/1572] Compiling C++ object scipy/stats/_boost/hypergeom_ufunc.cpython-311-x86_64-linux-gnu.so.p/meson-generated_hypergeom_ufunc.cpp.o
>       [800/1572] Compiling C++ object scipy/stats/_boost/nbinom_ufunc.cpython-311-x86_64-linux-gnu.so.p/meson-generated_nbinom_ufunc.cpp.o
>       [801/1572] Compiling C++ object scipy/stats/_boost/binom_ufunc.cpython-311-x86_64-linux-gnu.so.p/meson-generated_binom_ufunc.cpp.o
>       [802/1572] Compiling C++ object scipy/stats/_boost/ncf_ufunc.cpython-311-x86_64-linux-gnu.so.p/meson-generated_ncf_ufunc.cpp.o
>       [803/1572] Compiling C++ object scipy/stats/_boost/beta_ufunc.cpython-311-x86_64-linux-gnu.so.p/meson-generated_beta_ufunc.cpp.o
>       [804/1572] Compiling C object scipy/stats/_stats.cpython-311-x86_64-linux-gnu.so.p/meson-generated__stats.c.o
>       [805/1572] Compiling C object scipy/special/cython_special.cpython-311-x86_64-linux-gnu.so.p/meson-generated_cython_special.c.o
>       [806/1572] Compiling C object scipy/linalg/_flapack.cpython-311-x86_64-linux-gnu.so.p/meson-generated_..__flapackmodule.c.o
>       [807/1572] Compiling C++ object scipy/sparse/sparsetools/_sparsetools.cpython-311-x86_64-linux-gnu.so.p/csr.cxx.o
>       [808/1572] Compiling C++ object scipy/sparse/sparsetools/_sparsetools.cpython-311-x86_64-linux-gnu.so.p/bsr.cxx.o
>       ninja: build stopped: subcommand failed.
>       + meson setup --native-file=/tmp/pip-install-ur69mgkt/scipy_2cd8b05e22454fca908c9519705c3b1f/.mesonpy-native-file.ini -Ddebug=false -Doptimization=2 --prefix=/usr/local /tmp/pip-install-ur69mgkt/scipy_2cd8b05e22454fca908c9519705c3b1f /tmp/pip-install-ur69mgkt/scipy_2cd8b05e22454fca908c9519705c3b1f/.mesonpy-5ed401on/build
>       + meson compile
>       Traceback (most recent call last):
>         File "/home/sparky4/.local/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
>           main()
>         File "/home/sparky4/.local/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
>           json_out['return_val'] = hook(**hook_input['kwargs'])
>                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/home/sparky4/.local/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
>           return hook(config_settings)
>                  ^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 970, in get_requires_for_build_wheel
>           if not project.is_pure and platform.system() == 'Linux':
>                  ^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 875, in is_pure
>           return bool(self._wheel_builder.is_pure)
>                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/usr/local/lib/python3.11/functools.py", line 1001, in __get__
>           val = self.func(instance)
>                 ^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 209, in is_pure
>           if self._wheel_files['platlib']:
>              ^^^^^^^^^^^^^^^^^
>         File "/usr/local/lib/python3.11/functools.py", line 1001, in __get__
>           val = self.func(instance)
>                 ^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 170, in _wheel_files
>           return self._map_to_wheel(self._sources, self._copy_files)
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 512, in _map_to_wheel
>           or self._map_from_heuristics(
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 476, in _map_from_heuristics
>           self._warn_unsure_platlib(origin, destination)
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 438, in _warn_unsure_platlib
>           if self._is_native(origin):
>              ^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 405, in _is_native
>           self._project.build()  # the project needs to be built for this :/
>           ^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 765, in build
>           self._meson('compile')
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 696, in _meson
>           return self._proc('meson', *args)
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^
>         File "/tmp/pip-build-env-9_w94buz/overlay/lib/python3.11/site-packages/mesonpy/__init__.py", line 691, in _proc
>           subprocess.check_call(list(args))
>         File "/usr/local/lib/python3.11/subprocess.py", line 413, in check_call
>           raise CalledProcessError(retcode, cmd)
>       subprocess.CalledProcessError: Command '['meson', 'compile']' returned non-zero exit status 1.
>       [end of output]
>   
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: subprocess-exited-with-error
> 
> × Getting requirements to build wheel did not run successfully.
> │ exit code: 1
> ╰─> See above for output.
> 
> note: This error originates from a subprocess, and is likely not a problem with pip.
> ```

I get the same `TypeError: 'module' object is not callable` error when pip installing scipy>=1.7.3 in python:3.11-slim

**Comment by andyfaff** [comment]
With an RPi5 I have no install problems installing a pre-built wheel from PyPI into a venv made with the system python.

**Comment by rgommers** [comment]
This is just starting to be an amalgamation of different problems. So I'll give some explanations and will then close the issue. For new build issues we'll need a new issue with a _complete_ description of the build setup, invocation, log file, etc.

And +1 to what @andyfaff just said: please use pre-built wheels when you can. For `aarch64` you can install from PyPI, and for 32-bit ARM you can install them from https://piwheels.org/.

> `ninja: build stopped: subcommand failed.` on `sparsetools_bsr.cxx`

This is the compiler running out of memory. Those `sparsetools` files take a lot of memory and a Raspberry Pi doesn't have that much.

> `../scipy/sparse/linalg/_propack/meson.build:69:29: ERROR: Can not run test applications in this cross environment.`

That's a cross compilation issue. You currently either need to run the build under QEMU, or some other workaround to ensure you don't try to run executables meant for a native machine. Don't try this unless you know what you are doing.
